### PR TITLE
Fix some models cache initialization

### PR DESCRIPTION
**Pull request by albertvillanova**
Fix Gemma3 cache initialization.

Current code:
```python
if use_cache and past_key_values is None and not self.training:
    past_key_values = DynamicCache(config=self.config)
```

The `not self.training` check prevents cache creation during training. However, GRPO (and similar RL trainers) call `model.generate()` during training to produce completions for scoring. This condition blocks cache creation, resulting in:
  - First forward: `past_key_values=None`, no cache → no error (prefill processes all tokens)
  - Second forward: `past_key_values=None`, `cache_position=[279]` → mask expects 280 tokens but KV has only 1 → **shape mismatch error**

This PR removes the `not self.training` condition. 
```python
if use_cache and past_key_values is None:
    past_key_values = DynamicCache(config=self.config)
```
Now, the cache flow is:
  - First forward:  `past_key_values=None` → DynamicCache(length=279)
  - Second forward: `past_key_values=DynamicCache(279)` → DynamicCache(length=280)

Related issue in TRL:
- https://github.com/huggingface/trl/issues/4593

CC: @Cyrilvallez and @vasqu, who were investigating the underlying issue

### CODE REVIEW

**Comment by Cyrilvallez** [comment]
I will check more, but I believe the true issue comes from the line I showed in my comment https://github.com/huggingface/trl/issues/4593#issuecomment-3606035594! Here, you're only treating a symptom I think, not the full problem. Also, you said in the issue that you were not in training mode 👀

Usually, even in the middle of training, you will go back to inference mode to perform evaluation in order to deactivate dropout etc, unless I'm very mistaken

**Comment by Cyrilvallez** [comment]
Actually I was wrong, and this is the correct fix @albertvillanova! 

@vasqu because the input preparation relies on `use_cache` and not on `cache is None` to know whether we'll have a cache or not, there was a silent bug in which the input would be sliced as if a cache was used, but due to the `is training` conditional, no cache was provided. So generation would be successful, but completely wrong, as it would just see the last token in each generation step. The change in `mask_utils` actually uncovered it, because it shown the mismatch between what the size should be and what the size actually is, because we should have a cache but don't.... Before we would just prepare a mask of size 1, which is the input size, but the input size should NOT be that.
It started being an issue since https://github.com/huggingface/transformers/pull/41505 2 months ago... Before that, the Cache would have been there anyway and no problem

@albertvillanova I'm going to push much more changes on this PR, as we should really be more careful about this!
In the future, I'm going to make sure that `generate` relies on whether we actually have a cache or not, and not on `use_cache` which can be decorrelated from the cache existence as it's the case here

**Comment by vasqu** [comment]
> Usually, even in the middle of training, you will go back to inference mode to perform evaluation in order to deactivate dropout etc, unless I'm very mistaken

I think we are lucky here that most newer models do not use dropout so there is no real difference in inference. I agree tho, in general, the model should be set to eval `model.eval()` just to be sure.

> Here, you're only treating a symptom I think, not the full problem. Also, you said in the issue that you were not in training mode 👀

I think this really is the core issue --> We should not cut our inputs at all when we do not use a cache (as indicated by the config flag). However, the default flag is set to to `True`. This forces the cutting of the input, except you explicitly pass `use_cache=False` yourself. That shouldn't be the case at all! 

Some quick pointers what happens:
- https://github.com/huggingface/transformers/blob/75c135d76f6db226d83213c6bb658a474dddb07d/src/transformers/generation/utils.py#L616 We get `use_cache=True` here
- https://github.com/huggingface/transformers/blob/75c135d76f6db226d83213c6bb658a474dddb07d/src/transformers/generation/utils.py#L619-L625 We get into the cache-dependent input prep
- https://github.com/huggingface/transformers/blob/75c135d76f6db226d83213c6bb658a474dddb07d/src/transformers/generation/utils.py#L523 Our input gets cut (even if it shouldn't

The real solution IMO is as @Cyrilvallez said: Remove the default for `use_cache`, e.g. by removing it completely there and let it be handled within the kwargs.

**Comment by vasqu** [comment]
Ok, but then let's do both @Cyrilvallez: This fix along the kwarg being defaulted to `True`. It's not good either way

**Comment by Cyrilvallez** [comment]
Haha we commented at the same time

Yeah agree the kwargs should not be True, but it's not the one at fault here

File: src/transformers/models/cohere2/modular_cohere2.py [hunk 1]
```diff
@@ -367,7 +367,7 @@ def forward(
         if inputs_embeds is None:
             inputs_embeds = self.embed_tokens(input_ids)
 
-        if use_cache and past_key_values is None and not self.training:
+        if use_cache and past_key_values:
```
Comment (vasqu) [inline src/transformers/models/cohere2/modular_cohere2.py]:
```suggestion
        if use_cache and past_key_values is None:
```
Careful here

File: src/transformers/models/fast_vlm/modeling_fast_vlm.py
Comment (vasqu) [inline src/transformers/models/fast_vlm/modeling_fast_vlm.py:1]: Should be handled on main iirc

**Review by vasqu — COMMENTED** [review]
Small intermediate review

File: src/transformers/models/fast_vlm/modeling_fast_vlm.py
Comment (Cyrilvallez) [reply src/transformers/models/fast_vlm/modeling_fast_vlm.py:1]: Yep indeed just noticed it and rebased

File: src/transformers/models/gemma3/modular_gemma3.py [hunk 2]
```diff
@@ -934,6 +881,7 @@ class Gemma3ForConditionalGeneration(PaliGemmaForConditionalGeneration):
     # Fix: https://github.com/huggingface/transformers/issues/40564
     accepts_loss_kwargs = False
 
+    @can_return_tuple
```
Comment (albertvillanova) [inline src/transformers/models/gemma3/modular_gemma3.py:883]: I think, after your changes, multiple `forward` methods cannot return a tuple anymore. Their return signature should be updated as well.

```suggestion
```

File: src/transformers/models/gemma3n/modeling_gemma3n.py [hunk 3]
```diff
@@ -1426,154 +1420,7 @@ def forward(
...
-
-    def forward(self, x):
-        down_proj = self.down_proj(self.act_fn(self.gate_proj(x)) * self.up_proj(x))
-        return down_proj
-
-
-class Gemma3nAttention(nn.Module):
-    """Multi-headed attention from 'Attention Is All You Need' paper"""
-
-    def __init__(self, config: Gemma3nConfig, layer_idx: int):
```
Comment (Cyrilvallez) [inline src/transformers/models/gemma3n/modeling_gemma3n.py:1456]: Those classes should never have been here... Only the `Text` ones are needed. Because `_can_record_outputs` was not correctly overwritten with the correct `Text` prefix, they were imported...

File: src/transformers/models/gemma3/modular_gemma3.py (hunk 2 above)
Comment (vasqu) [reply src/transformers/models/gemma3/modular_gemma3.py:883]: Hmm, not sure if I can follow:
- `can_return_tuple` allows users to return a tuple if they pass `return_dict=False`
- Defaults to returning the output class tho

After multiple forwards it still depends on how you call it.

Comment (Cyrilvallez) [reply src/transformers/models/gemma3/modular_gemma3.py:883]: The decorator automatically handle tuple casting if needed!

File: src/transformers/models/gemma3/modular_gemma3.py [hunk 4]
```diff
@@ -527,23 +527,19 @@ def forward(
         attention_mask: Optional[torch.Tensor] = None,
         position_ids: Optional[torch.LongTensor] = None,
         past_key_values: Optional[Cache] = None,
-        output_attentions: Optional[bool] = False,
-        use_cache: Optional[bool] = False,
         cache_position: Optional[torch.LongTensor] = None,
         **kwargs,
```
Comment (vasqu) [inline src/transformers/models/gemma3/modular_gemma3.py]: Should we add the `TransformersKwargs` here for typing, just for completeness sake (and elsewhere)

**Comment by Cyrilvallez** [comment]
Pushed the necessary change related to this PR on several models, and took the opportunity to do quite some cleanup at the same time!

File: src/transformers/models/gemma3n/modeling_gemma3n.py (hunk 3 above)
Comment (vasqu) [reply src/transformers/models/gemma3n/modeling_gemma3n.py:1456]: Wow, so the manual collecting was needed because we didn't properly type out out records either way

File: src/transformers/models/gemma3n/modular_gemma3n.py [hunk 5]
```diff
@@ -1976,7 +1972,8 @@ def project_per_layer_inputs(
             dtype=inputs_embeds.dtype, device=per_layer_projection.device
         )
 
-    @can_return_tuple
+    # Last hidden states should be before reprojecting it to stay consistent, so we do not overwrite with the final one
```
Comment (vasqu) [inline src/transformers/models/gemma3n/modular_gemma3n.py]:
```suggestion
    # Last hidden states should be before reprojecting it to stay consistent with layer output
```
Just nit, but felt like shortening it

File: src/transformers/models/gemma3n/modular_gemma3n.py [hunk 6]
```diff
@@ -1987,46 +1984,28 @@ def forward(
...
             past_key_values = DynamicCache(config=self.config)
 
         if cache_position is None:
             past_seen_tokens = past_key_values.get_seq_length() if past_key_values is not None else 0
-            cache_position = torch.arange(
-                past_seen_tokens,
-                past_seen_tokens + inputs_embeds.shape[1],
-                device=inputs_embeds.device,
-            )
+            cache_position = torch.arange(inputs_embeds.shape[1], device=inputs_embeds.device) + past_seen_tokens
```
Comment (vasqu) [inline src/transformers/models/gemma3n/modular_gemma3n.py:2007]: Hmm, that was to avoid a device sync iirc? Just fmi

**Review by vasqu — APPROVED** [review]
LGTM, just some smaller comments + let's just make sure to check with slow tests that nothing has been significantly broken

File: src/transformers/models/gemma3n/modular_gemma3n.py (hunk 6 above)
Comment (Cyrilvallez) [reply src/transformers/models/gemma3n/modular_gemma3n.py:2007]: Yeah apparently `arange` sync by calling `.item` if a tensor on-device is passed to it.... 🫠🫠🙃🙃
